### PR TITLE
feat(rpc): implement debug_executionWitness API

### PR DESCRIPTION
**Pull request by 0x00101010**
**Description**

implement debug_executionWitness API

**Tests**

Tested offline with reth and op-program

**Additional context**

**Metadata**

### CODE REVIEW

**Comment by ajsutton** [comment]
Adding myself as reviewer so I don't forget this after the morning meeting run but no need to block on me if someone else gets to this first.

File: core/stateless/encoding.go [hunk 1]
```diff
@@ -74,3 +76,33 @@ type extWitness struct {
 	Codes   [][]byte
 	State   [][]byte
 }
+
+// ExecutionWitness is a witness json encoding for transferring across clients
+// in the future, we'll probably consider using the extWitness format instead for less overhead.
+// currently we're using this format for compatibility with the reth and also for simplicify in terms of parsing.
+type ExecutionWitness struct {
```
Comment (ajsutton) [inline core/stateless/encoding.go]: nit: 
```suggestion
// currently we're using this format for compatibility with reth and also for simplicity in terms of parsing.
type ExecutionWitness struct {
```

File: eth/api_debug.go [hunk 2]
```diff
@@ -446,3 +447,37 @@ func (api *DebugAPI) GetTrieFlushInterval() (string, error) {
 	}
 	return api.eth.blockchain.GetTrieFlushInterval().String(), nil
 }
+
+func (api *DebugAPI) ExecutionWitness(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) (*stateless.ExecutionWitness, error) {
```
Comment (ajsutton) [inline eth/api_debug.go:452]: We should add a test for this as well. The RPC tests aren't particularly comprehensive but there is typically one for each API that hits on the key functionality.

File: eth/api_debug.go [hunk 3]
```diff
@@ -446,3 +447,37 @@ func (api *DebugAPI) GetTrieFlushInterval() (string, error) {
...
+		return nil, fmt.Errorf("failed to create witness: %w", err)
+	}
+
+	parentHeader := witness.Headers[0]
+	statedb, err := api.eth.blockchain.StateAt(parentHeader.Root)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve parent state: %w", err)
+	}
+
+	statedb.StartPrefetcher("debug_execution_witness", witness)
```
Comment (ajsutton) [inline eth/api_debug.go:477]: Seems like we should have a `defer StopPrefetcher()` so we don't leave it running?

**Review by ajsutton — COMMENTED** [review]
Looks good generally but would be worth adding a test for it.  @refcell  or @clabby does the returned format match what you'd expect?

There's also a comment saying the format might be simplified in the future - should be consider doing that now so we don't have to break compatibility? I'm not too sure what's actually likely to get standardised here.

File: eth/api_debug.go (hunk 3 above)
Comment (0x00101010) [reply eth/api_debug.go:477]: Great point!

File: eth/api_debug.go (hunk 2 above)
Comment (0x00101010) [reply eth/api_debug.go:452]: Cool, I'll try to add one that tests re-execution from witness

**Comment by 0x00101010** [comment]
> Looks good generally but would be worth adding a test for it. @refcell or @clabby does the returned format match what you'd expect?
> 
> There's also a comment saying the format might be simplified in the future - should be consider doing that now so we don't have to break compatibility? I'm not too sure what's actually likely to get standardised here.

regarding format, I believe so, I compared returns from reth & geth, and also the queries from op-program, they match.

Currently trying to figure out return mismatch between reth & geth, working with reth teams on that

**Comment by 0x00101010** [comment]
> > Looks good generally but would be worth adding a test for it. @refcell or @clabby does the returned format match what you'd expect?
> > There's also a comment saying the format might be simplified in the future - should be consider doing that now so we don't have to break compatibility? I'm not too sure what's actually likely to get standardised here.
> 
> regarding format, I believe so, I compared returns from reth & geth, and also the queries from op-program, they match.
> 
> Currently trying to figure out return mismatch between reth & geth, working with reth teams on that

@ajsutton, also regarding standardization, currently the conversation is around two things:
1. what exact data do we want to return here, geth returns headers in execution witness as well, while reth does not (currently not needed by op-program) => this is likely to be added into reth if we want the witness to be helpful to be used to execute blocks directly
2. how we want to return the data, from Peter's proposal, he suggests rlp everything to improve performance, currently since performance isn't a big concern, we opted for simplicify which is directly return json data, that can be revisited if needed

Added tests and ready for another review

**Review by ajsutton — APPROVED** [review]
LGTM. Given the benefit to fault proofs I think this is worth having in op-geth now, but would be good to submit this to upstream as well since it's a generically useful API.

**Comment by 0x00101010** [comment]
> LGTM. Given the benefit to fault proofs I think this is worth having in op-geth now, but would be good to submit this to upstream as well since it's a generically useful API.

Thanks @ajsutton, haven't done it now because there's still unanswered questions regarding return format and I sense resistance from geth side, but will try it to see how geth team reacts

**Comment by ajsutton** [comment]
Yeah, I'd say it's worth creating the PR to upstream to push the conversation forward.

**Pull request merged**